Fix image centering

McTalian opened the pull request:
Rendering at any given size would result in the identicon being
off-center. Rendering it at 300 and downsizing it seems to provide more
centered icons. Still allows users to specify natural size.

Commit:
Fix image centering

Rendering at any given size would result in the identicon being
off-center. Rendering it at 300 and downsizing it seems to provide more
centered icons. Still allows users to specify natural size.

kether667 commented:
is 300 the lowest that they appear centered? I'm a little concerned with performance when that would be used on like message list screen where you repeat it multiple times.

i mean you provided means to change it i'm just looking for sensible default

Mctalian commented:
I was also a bit concerned. I know Github's stored images are 420px, but
your demo's max was 300 so I assumed there was a reason for that.

Either way, I have an idea to make this more efficient. I'll get back to
you with an update.
On Nov 19, 2015 1:54 PM, "Lukasz Karolewski" <notifications@github.com>
wrote:

> is 300 the lowest that they appear centered? I'm a little concerned with
> performance when that would be used on like message list screen where you
> repeat it multiple times.
>
> i mean you provided means to change it i'm just looking for sensible
> default
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kether667/angular-identicon/pull/2#issuecomment-158154625>
> .
>

Commit:
Picks next best natural size based on size

User can still provide their own, but it calculates it if it is not
provided. The pattern for near perfect identicons starting with 10 are:
10, 17, 22, 29, 34, 41, etc. So every 2 sizes are 12 apart. I add 12
until we pass the provided $scope.size and subtract 5 to try to use the
lower value if we can.

The lowest natural size this is configured for is 17px. Theoretically if
someone wanted one smaller than 10px, 10px would be a better natural
size, but I assumed more people would be in the range of 24px minimum so
it just cuts off an extra loop just in case (likely not even needed).
But either way, I believe 15px is the minimum for identicons to even be
useful.

Also added one test in case you ever set up the framework (it doesn't
appear to be set up right now).

Mctalian commented:
Is my update a better solution?

kether667 commented:
when i opened Identicon source code i've notice there is a 3 param in constructor, `margin`

Identicon = function(hash, size, margin){
        this.hash   = hash;
        this.size   = size   || 64;
        this.margin = margin || .08;
    }

can you play around with margin, that for loop is not the best approach, there should one one time calculation

Mctalian commented:
I will look into this over the next few days and hopefully come up with a better solution.

Commit:
Fix identicon centering by calculating a margin

Calculate an appropriate margin for the identicon based on the provided
image size.

Mctalian commented:
Ok, so this isn't perfect, yet, but I wanted you to take a look at the approach before I sunk more time into it.

We calculate the margin based on the provided size.

In the end, I'd like to shoot for a margin that is ~25% of the cell size (`$scope.size / 5`) but still maintains the centering. I think that's an aesthetic margin size. Small identicons don't have enough space to make the image centered no matter what margin is provided or calculated, because we can't have fractions of pixels.

There are some optimizations for certain sizes that can be made. Give me a thumbs up on this and I'll look into making those.

kether667 commented:
yea i think you're on the right track

Commit:
Combination of margin and natural size methods

Images will be centered all the time with a combination of the margin
and natural size methods. At most, image sizes provided by the user will
be increase by 5px then scaled back down to the user's provided size.

Mctalian commented:
Alright, sorry it took so long! I decided to merge the naturalSize method with the margin method to make the image always centered but without the nasty for-loop in the original. I think this is much more efficient considering at most, it generates an image with a size that is 5 greater than the user's requested size and scales it down.

If a user elects to pass in a margin in pixels (marginPx), then we won't calculate anything and we will allow their provided margin size to take over.

Mctalian commented:
@kether667 is this solution agreeable?

kether667 commented:
yes it looks good, let me do some testing and will get back to you probably next monday

Mctalian commented:
Happy new year! Did you get a chance to test it?

Mctalian commented:
@kether667 friendly reminder bump

Mctalian commented:
Thanks for merging it! Can you generate the build files in dist and bump the version as well?
